FullStorySegmentMiddleware/Classes/FullStoryMiddleware.m

```diff
@@ -121,12 +127,168 @@ - (SEGPayload *) getNewPayloadWithFSURL:(SEGContext * _Nonnull)context {
+        } else if ([@"B" isEqualToString:@(type)]) {
+            // bool-type gets encoded as number number, and doesn't get encoded into 'B', but check anyway
+            suffix = @"_bool";
+        }
+    } else if ([obj isKindOfClass:[NSDate class]]) {
+        suffix = @"_date";
+    } else if([obj isKindOfClass:[NSString class]]) {
+        suffix = @"_str";
+    }
+
+    return suffix;
+}
+
+- (NSDictionary *)getDictionaryFromArrayObject:(NSArray *)arr withKey:(NSString *)key {
+    // FS does not accept array of dicts, only "premitive" arrays allowed
+    // So if it's an array of dicts or nested arrays, then we need to convert the root array to a dict so it becomes nested dicts
+    NSMutableDictionary *dict = [[NSMutableDictionary alloc] init];
+
+    for (int i = 0; i < [arr count]; i++) {
+        NSObject *item = arr[i];
+
+        if ([item isKindOfClass:[NSDictionary class]]) {
+            // if array of dicts, we then loop through all dicts, flatten out each key/val into arrays, we will loose the object association but it allows user to search for each key/val in the array in FS (i.e. searching for one product when array of products are sent)
+            NSMutableDictionary *tempDict = [[NSMutableDictionary alloc] initWithDictionary:[self getSuffixedProps:(NSDictionary *) item]];
+            [self appendToDictionary:dict withKey:key andValue:tempDict];
+        } else if ([item isKindOfClass:[NSArray class]]) {
+            // TODO: Segment spec should not allow nested array properties, ignore for now, but we should handle it eventually
+            NSDictionary *tempDict = [self getDictionaryFromArrayObject:(NSArray *)item withKey:key];
+            [self appendToDictionary:dict withKey:@"" andValue:tempDict];
+        } else {
+            // default to simple object
+            // if there are arrays of mixed type, then in the final props we will add approporate values to the same, key but with each type suffix
+            NSString* suffix = [[self getSuffixStringFromSimpleObject:item] stringByAppendingString:@"s"];
+            // get the current array form this specified suffix, if any, then append current item
+            NSMutableArray *tempArr = [[NSMutableArray alloc] initWithArray:[dict valueForKey:[key stringByAppendingString:suffix]]];
+            [tempArr addObject:item];
+            [dict setValue:tempArr forKey:[key stringByAppendingString:suffix]];
+        }
+    }
+
+    return dict;
+}
+
+- (void)appendToDictionary:(NSMutableDictionary *)dict withKey:(NSString *)key andDictionary:(NSDictionary *)dict2 {
+    // when adding a parsed dict into the result dict, emurate and add each object
+    for (NSString *k in dict2) {
+        NSString *nestedKey;
+        if ([key length] > 0) {
+            nestedKey = [[key stringByAppendingString:@"."] stringByAppendingString:k];
+        } else {
+            nestedKey = k;
+        }
+        [self appendToDictionary:dict withKey:nestedKey andValue:dict2[k]];
```

Comment:
^recursion works for now since Segment should not have deeply nested stuff nor circular dependency, but should refactor to bfs with max depth at some point